### ISSUE TITLE
Navigating with GPX no max.speed / speed camera warning

### ISSUE THREAD

**Issue by Alootje** [osmandapp/Osmand#2843]
After creating a multipoint route in OsmAnd and start navigating the max. speed is shown and (for instance) the app warns for speed cameras.
If I save this multipoint route into a GPX and use the GPX file for navigating that the max. speed and speed camera notifications are not shown.

- Version of OsmAnd: 2.3.4
- Android phone (Samsung galaxy Note 3, N9005)
- With offline maps offered within the OsmAnd app for download, using OsmAnd's in-app offline routing.
- Selected profile: motorbike

**Comment by sonora** [comment]
@vshcherb Confirmed. Weird - we could simply implement a workaround to detect if GPX routing is in use and then derive the traffic warnings from extrapolating the last known position  (like we do for the `isFollowingMode()=false` case) rather than from the current route segment.

But I leave this untouched for now for Victor to double-check why this does not natively work with GPX routing in the first place.

**Comment by scaidermern** [comment]
I also noticed some while ago. Similarly, turn arrows are not displayed. This only happens when using a GPX track and not when using a GPX route.

**Comment by vshcherb** [comment]
@sonora yes we could, I think RoutingHelper.getCurrentGPXRoute() != null should be a good indicator

**Comment by sonora** [comment]
https://github.com/osmandapp/Osmand/commit/c96a3976c6825c221e4c99c3890f258276d3f6c5 - still needs testing.

**Comment by Alootje** [comment]
@sonora, thank you for the update. How can I assist in testing this?

**Comment by sonora** [comment]
It is in our nightly build, check http://osmand.net, you could install and test it.

Depending on your setup you may or may not be able to commonly use the maps you have downloaded already for both your release and the nightly build. I am actually not sure if we have any useful documentation on this question.

**Comment by Alootje** [comment]
Thought I had to exchange some .Java files myself, but installing the nightly build is more convenient :-)

I tested the nightly build (with route simulation, not actual driving). Create a from A to B route, with one via point. Saved is as GPX and used that for GPX for the simulation. The maximum speed and speed cameras seem to be displayed, see screenshots below.

It does however seem that the speed camera icon and audible notification keeps repeating. On the screenshot (1) below the speed camera is already passed...

Furthermore the audible notifications on the roundabouts seems to be correct. I thought I read some mistakes happen here when navigating with a GPX. During my simulation run the lady said to 'take the second exit'.

Final thing I noticed was that the duration differs a lot. While creating the route OsmAnd stated the route is 17.x km and 23 minutes. Selecting the GPX came a with a similar length, however the duration was only 13 minutes....

1. S3_GPX_V2.4.0_4202M
![1 s3_gpx_v2 4 0_4202m](https://cloud.githubusercontent.com/assets/20538605/17135033/ee9c1f6a-532d-11e6-996b-a2518238e9a5.png)

2. Note3_GPX_v2.3.4
![2 note3_gpx_v2 3 4](https://cloud.githubusercontent.com/assets/20538605/17135043/fdb2901a-532d-11e6-8f70-81088e18024c.png)

**Comment by scaidermern** [comment]
Working maxspeed is great to hear.

The estimated duration / arrival time have always been wrong when using GPX tracks for navigation (is there already an issue for this?). According to your first screenshot the issue with turn arrows being blocked by the progress bar (#2592) is also still reproducible.

**Comment by sonora** [comment]
Ok, thanks! Would be good to verify all findings in the "real world".

**(1) Hearing traffic warnings twice, before and after passing them:** May be an unrelated bug. Sounds like our code to filter only traffic warnings in "forward" direction is buggy, I think I may have seen similar symptoms in other situations.

**(2) Roundabouts:** Did you mean to say all is ok, or report an issue? There may be a word missing in your paragraph, and the respective screenshot?

**(3) Travel time:** You can toggle between "arrival time" and "travel time left" by tapping on the time widget. And in deed, your screenshots seem to indicate 11 vs. 17  minutes left, which sounds like a significant difference. How does it look when creating both routes with the **same** app version?

**Comment by Alootje** [comment]
(0) my "real world" testing will be earliest coming weekend, maybe even the weekend after that. 

(1) Is this reported by an issue? If not, does it required an issue? 

(2) TLDR: thought this was an issue, but it wasn't.
I **_thought_** that in V2.3.4/V2.3.5 the exit number wasn't announced. During simulation mode on v2.4.0_4202 the lady announced 'enter the roundabout and take the second exit', which is a good thing. Checked this on V2.3.5. as well, there the lady announces the same thing.
One thing I did notice was that the arrows at the roundabout are different: the OsmAnd created route shows the arrow up and until one quarter/third of the roundabout and the GPX route shows the arrow until the roundabout. Don't know if this affects navigation? This is in both v2.3.5 (screenshots) as v2.4.0_4202.

A. Roundabout_arrow_OsmAnd_route
![roundabout_arrow_osmand_route](https://cloud.githubusercontent.com/assets/20538605/17142712/6a2844ba-5351-11e6-8167-cc615f34a203.png)
B. Roundabout_arrow_GPX
![roundabout_arrow_gpx](https://cloud.githubusercontent.com/assets/20538605/17142724/789a7c5c-5351-11e6-87e7-aa248f334287.png)

(3) I know about the toggle between "arrival time" and "travel time left". Below a comparison between both versions (performed on a Samsung galaxy S3, however similar results on my Note 3):
**V2.3.5**_Route_created_in_OsmAnd:
- 17.6 km
- 24 min

**V2.3.5**_GPX_from_above (v2.3.5):
- 17.6 km
- 13 min

**V2.4.0_4202**_Route_created_in_OsmAnd:
- 17.5 km
- 23 min

**V2.4.0_4202**_GPX_from_above (V2.4.0_4202):
- 17.5 km
- 12 min

I think item 3 requires a separate issue?

(4) While navigating with a GPX the lane guidance (middle section at the top of the screen doesn't work - great feature btw! Didn't know of its existence until just now). The first screenshot shows navigating a route created in OsmAnd 2.4.0_4202. The second and third screenshot show navigating the GPX of that same route.

C. Lane Guidance v2.4.0_4202 OsmAnd Route 
![v2 4 0_4202_osmand_route](https://cloud.githubusercontent.com/assets/20538605/17142856/e2e81e70-5351-11e6-9cf4-9fbdacada284.png)
D. Lane Guidance v2.4.0_4202 GPX (1 fof 2)
![v2 4 0_4202_gpx_1](https://cloud.githubusercontent.com/assets/20538605/17142863/e7b33bba-5351-11e6-8e93-bc5919eaa553.png)
E. Lane Guidance v2.4.0_4202 GPX (2 fof 2)
![v2 4 0_4202_gpx_2](https://cloud.githubusercontent.com/assets/20538605/17142868/eb73274c-5351-11e6-837b-6738c1aea5fd.png)

**Comment by sonora** [comment]
(1) Yes separate Issue may be helpful I had briefly looked at our code and think we apply directional analysis so far only to speed limits, not yet to any other traffic warnings (when based on location and not routing data)

(2) Many changes have happened since 2.3.4, to voice prompting and also to the roundabout drawing, see also Issue #2745. Navigation is independent code, though.

(3) Arrival time for GPX routing probably worth a separate issue.

(4) Yes, I had applied my fix also to the turn lanes widget, but do not know how well it works there.

In general, I would say that GPX routing should mainly be applied where not alternative is available, like when you need to follow a prepared tour, like from a guide book. In all other cases it is our goal that OsmAnd's offline routing engine should provide the better optimized an more dynamic solution.

**Comment by Alootje** [comment]
(1) I shall create an issue within the next days.

(2) In my opinion the voice guidance works as it should (during route simulation at least). I didn't notice the difference in the roundabout icons explained in issue #2745.

(3) I shall create an issue for this within the next days as well.

(4) As can be seen in the screenshots in my previous post the turn lanes widget doesn't work as it does during 'normal navigation'.

I agree that navigating via GPX seems inferior to OsmAnds offline routing capabilities, however I like to create routes in OsmAnd on day X and ride them on day Y. What I also find convenient about GPX is that I do not have to follow the entire track (if this option is not selected). If for instance a road is closed I can take a detour and continue the track when I get back on it. This without action from the user. Via points on a 'OsmAnd created route' require to be passed. If one is missed than OsmAnd will send you back to this point.
I've found some things and experimented with the new 'Map markers' feature which, for me this is not convenient.
If I could save a 'OsmAnd created route' (start-, via- and end points) to a, for instance .osm file and load it later on would be a nice solution. Maybe in addition with a 'Skip next waypoint' button? Exporting the route to .GPX becomes a nice to have for sharing to the Garmin fellas who join me ;-).

This all is from a personal perspective. If these features (or some of them) are already available in OsmAnd please point me in the right direction (haha) and I will experiment with it.

**Comment by sonora** [comment]
Underlying issue fixed, collateral findings broken out to specific issues.
